src/OrchardCore/OrchardCore.ResourceManagement/FileVersionProvider.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Antiforgery.Internal;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.FileProviders;
+using OrchardCore.Abstractions.Modules.FileProviders;
+
+namespace OrchardCore.ResourceManagement
+{
+    /// <summary>
+    /// Provides version hash for a specified file.
+    /// </summary>
+    public class FileVersionProvider : IFileVersionProvider
+    {
+        private const string VersionKey = "v";
+        private static readonly char[] QueryStringAndFragmentTokens = new[] { '?', '#' };
+
+        private readonly IEnumerable<IFileProvider> _registeredFileProviders;
+        private readonly IFileProvider _webRootfileProvider;
+
+        private IList<IFileProvider> _combinedFileProviders;
+        private readonly IMemoryCache _cache;
+        public FileVersionProvider(
+            IEnumerable<IFileProvider> registeredFileProviders,
```

Comment:
So, on my side i use `IEnumerable<IStaticFileProvider> registeredFileProviders` just for the DI. But i still use `IEnumerable<IFileProvider> _registeredFileProviders`.

Then i don't use `_combinedFileProviders` and `_webRootfileProvider`, only `_registeredFileProviders` that i init in the ctor like this.

            _registeredFileProviders = registeredFileProviders
                .Concat(new[] { environment.WebRootFileProvider });

Also, `OC.Media` can be used without `OC.ResourceManagement`, and `OC.ResourceManagement` needs to be independent so that it can be used without `OC.Abstractions`. But `OC.ResourceManagement` needs mvc, so i moved this file to `OC.Mvc.Core` under the folder / namespace `OrchardCore.Mvc.FileProviders`. Then i register it in the `OrchardCore.Mvc.Startup`.

So, without `OC` the `ResourceManager` will use the default mvc version provider, with `OC` and our `AddMvc()` it will use our version provider.

Reply:
And with your other change to always create `wwwroot` no need to null check `environment.WebRootFileProvider`

Reply:
The `WebRootFileProvider` may be a `NullFileProvider` but it is never null.